packages/idempotency/src/persistence/DynamoDbPersistenceLayer.ts

```diff
@@ -1,18 +1,73 @@
 /* eslint-disable @typescript-eslint/no-empty-function */
+import { DynamoDB } from '@aws-sdk/client-dynamodb';
+import { DynamoDBDocument, GetCommandOutput } from '@aws-sdk/lib-dynamodb';
+import { IdempotencyItemNotFoundError } from '../Exceptions';
+import { IdempotencyRecordStatus } from '../types/IdempotencyRecordStatus';
 import { IdempotencyRecord, PersistenceLayer } from './PersistenceLayer';
 
 class DynamoDBPersistenceLayer extends PersistenceLayer {
-  public constructor(_tableName: string, _key_attr: string = 'id') {
+  private _table: DynamoDBDocument | undefined;
+
+  public constructor(private tableName: string, private key_attr: string = 'id',
+    private status_attr: string = 'status', private expiry_attr: string = 'expiration',
+    private in_progress_expiry_attr: string = 'in_progress_expiry_attr',
+    private data_attr: string = 'data') {
     super();
   }
-  protected async _deleteRecord(): Promise<void> {}
-  protected async _getRecord(): Promise<IdempotencyRecord> {
-    return Promise.resolve({} as IdempotencyRecord);
+
+  protected async _deleteRecord(record: IdempotencyRecord): Promise<void> {
+    const table: DynamoDBDocument = this.getTable();
+    await table.delete({
+      TableName: this.tableName, Key: { [this.key_attr]: record.idempotencyKey }
+    });
+  }
+
+  protected async _getRecord(idempotencyKey: string): Promise<IdempotencyRecord> {
+    const table: DynamoDBDocument = this.getTable();
+    const output: GetCommandOutput = await table.get(
+      {
+        TableName: this.tableName, Key: { [this.key_attr]: idempotencyKey }
+      }
+    );
+
+    if (!output.Item) {
+      throw new IdempotencyItemNotFoundError();
+    }
+
+    return new IdempotencyRecord(output.Item[this.key_attr], output.Item[this.status_attr], output.Item[this.expiry_attr], output.Item[this.in_progress_expiry_attr], output.Item[this.data_attr], undefined);
+  }
+
+  protected async _putRecord(_record: IdempotencyRecord): Promise<void> {
```

Comment:
Is it misleading to pass in an idempotency record into "putRecord" but then that record isn't actually saved as is? You override what keys are actually saved. Maybe the issue is just naming it "IdempotencyRecord" when it's not the record? idk

Reply:
A little confused on the question? I don't believe the consumer should ever look at the record. They will interact with this module by the errors it throws or the records it will eventually return.

Reply:
Fair enough, now that I think about it that makes sense. If they never see it, then it's less of an issue.